File: typescript/src/backends/dynamo.ts
import {
  ConditionalCheckFailedException,
  CreateTableCommand,
  DeleteTableCommand,
  type DynamoDBClient,
  ResourceInUseException,
} from "@aws-sdk/client-dynamodb";
import {
  DeleteCommand,
  GetCommand,
  PutCommand,
  UpdateCommand,
} from "@aws-sdk/lib-dynamodb";
import type { MemKey, Store } from "../store.ts";
import type { NativeAttributeValue } from "@aws-sdk/util-dynamodb";

import { setTimeout } from "node:timers/promises";

export class Dynamo implements Store {
  private readonly client: DynamoDBClient;
  private readonly tableName: string;

  public constructor(client: DynamoDBClient, tableName: string) {
    this.client = client;
    this.tableName = tableName;
  }

  public async has(key: MemKey): Promise<boolean> {
    const { Item } = await this.client.send(
      new GetCommand({
        TableName: this.tableName,
        Key: this.key(key),
        ProjectionExpression: "pk",
      }),
    );
    return !!Item;
  }

  public async get(key: MemKey): Promise<Uint8Array | undefined> {
    const { Item } = await this.runGet(key);
    return Item?.value;
  }

  public async getWithRetry(
    key: MemKey,
    maxRetries: number = 4,
    baseDelayMs: number = 25,
    factor: number = 2,
    maxBackoffMs: number = 300,
  ): Promise<Uint8Array | undefined> {
    let attempt = 0;
    while (true) {
      const { Item } = await this.runGet(key);
      if (Item?.value) return Item.value;

      if (attempt++ >= maxRetries) break;

      await setTimeout(Math.min(baseDelayMs * factor ** attempt, maxBackoffMs));
    }
  }

  public async set(key: MemKey, value: Uint8Array): Promise<void> {
    await this.client.send(
      new PutCommand({
        TableName: this.tableName,
        Item: {
          ...this.key(key),
          value,
        },
      }),
    );
  }

  public async delete(key: MemKey): Promise<void> {
    await this.client.send(
      new DeleteCommand({
        TableName: this.tableName,
        Key: this.key(key),
        ReturnValues: "ALL_OLD",
      }),
    );
  }

  public async setNewValue(key: MemKey, value: Uint8Array): Promise<boolean> {
    try {
      await this.client.send(
        new UpdateCommand({
          TableName: this.tableName,
          Key: this.key(key),
          UpdateExpression: "SET #value = :value",
          ConditionExpression: "attribute_not_exists(#value)",
          ExpressionAttributeNames: { "#value": "value" },
          ExpressionAttributeValues: { ":value": value },
        }),
      );
      return true;
    } catch (err: unknown) {
      if (err instanceof ConditionalCheckFailedException) {
        return false;
      }
      throw err;
    }
  }

  public async compareAndSet(
    key: MemKey,
    value: Uint8Array,
    expected: Uint8Array,
  ): Promise<boolean> {
    try {
      await this.client.send(
        new UpdateCommand({
          TableName: this.tableName,
          Key: this.key(key),
          UpdateExpression: "SET #value = :value",
          ConditionExpression: "#value = :expected",
          ExpressionAttributeNames: { "#value": "value" },
          ExpressionAttributeValues: {
            ":value": value,
            ":expected": expected,
          },
        }),
      );
      return true;
    } catch (err: unknown) {
      if (err instanceof ConditionalCheckFailedException) {
        return false;
      }
      throw err;
    }
  }

  public async compareAndDelete(
    key: MemKey,
    expected: Uint8Array,
  ): Promise<boolean> {
    try {
      await this.client.send(
        new DeleteCommand({
          TableName: this.tableName,
          Key: this.key(key),
          ConditionExpression:
            "attribute_exists(#value) AND #value = :expected",
          ExpressionAttributeNames: { "#value": "value" },
          ExpressionAttributeValues: { ":expected": expected },
        }),
      );
      return true;
    } catch (err: unknown) {
      if (err instanceof ConditionalCheckFailedException) {
        return false;
      }
      throw err;
    }
  }

  public async createTable(): Promise<void> {
    try {
      await this.client.send(
        new CreateTableCommand({
          TableName: this.tableName,
          KeySchema: [
            { AttributeName: "pk", KeyType: "HASH" },
            { AttributeName: "sk", KeyType: "RANGE" },
          ],
          AttributeDefinitions: [
            { AttributeName: "pk", AttributeType: "S" },
            { AttributeName: "sk", AttributeType: "S" },
          ],
          ProvisionedThroughput: {
            ReadCapacityUnits: 5,
            WriteCapacityUnits: 5,
          },
        }),
      );
    } catch (err: unknown) {
      if (err instanceof ResourceInUseException) {
        return;
      }
      throw err;
    }
  }

  public async deleteTable(): Promise<void> {
    await this.client.send(
      new DeleteTableCommand({
        TableName: this.tableName,
      }),
    );
  }

  private key(key: MemKey): Record<string, NativeAttributeValue> {
    return {
      pk: key.callHash,
      sk: key.type,
    };
  }

  private async runGet(key: MemKey) {
    return this.client.send(
      new GetCommand({ TableName: this.tableName, Key: this.key(key) }),
    );
  }
}

File: python/src/brrr/backends/dynamo.py
from __future__ import annotations

import asyncio
import functools
import logging
import typing
from typing import Any, Callable, Coroutine

from ..store import CompareMismatch, MemKey, NotFoundError, Store

if typing.TYPE_CHECKING:
    from types_aiobotocore_dynamodb import DynamoDBClient


logger = logging.getLogger(__name__)


# The frame table layout is:
#
#   pk: MEMO_KEY
#   sk: "pending_returns"
#   parents: list[str]
#
# OR
#
#   pk: MEMO_KEY
#   sk: "call"
#   task: The task name
#   argv: bytes (pickled)
#
# OR
#
#   pk: MEMO_KEY
#   sk: "value"
#   value: bytes (pickled)
#
# TODO It is possible we'll add versioning in there as pk or something


def async_retry_on_exception[**P, T](
    exception: type[Exception],
    max_retries: int,
    base_delay_ms: int,
    factor: int,
    max_backoff_ms: int,
) -> Callable[
    [Callable[P, Coroutine[Any, Any, T]]], Callable[P, Coroutine[Any, Any, T]]
]:
    def decorator(
        func: Callable[P, Coroutine[Any, Any, T]],
    ) -> Callable[P, Coroutine[Any, Any, T]]:
        @functools.wraps(func)
        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            retries = 0
            while True:
                try:
                    return await func(*args, **kwargs)
                except exception as e:
                    retries += 1
                    if retries > max_retries:
                        raise

                    logger.warning(
                        f"Retrying {func.__name__} after {e}, "
                        f"attempt {retries}/{max_retries}"
                    )

                    await asyncio.sleep(
                        min(base_delay_ms * (factor ** (retries)), max_backoff_ms)
                        / 1000
                    )

        return wrapper

    return decorator


class DynamoDbMemStore(Store):
    client: DynamoDBClient
    table_name: str

    def key(self, mem_key: MemKey) -> dict[str, dict[str, str]]:
        return {"pk": {"S": mem_key.call_hash}, "sk": {"S": mem_key.type}}

    def __init__(self, client: DynamoDBClient, table_name: str):
        self.client = client
        self.table_name = table_name

    async def has(self, key: MemKey) -> bool:
        return "Item" in await self.client.get_item(
            TableName=self.table_name,
            Key=self.key(key),
        )

    async def get(self, key: MemKey) -> bytes:
        response = await self.client.get_item(
            TableName=self.table_name,
            Key=self.key(key),
        )
        if "Item" not in response:
            logger.debug(f"getting key: {key}: not found")
            raise NotFoundError(key)
        logger.debug(f"getting key: {key}: found")
        return response["Item"]["value"]["B"]

    @async_retry_on_exception(
        exception=NotFoundError,
        max_retries=4,
        base_delay_ms=25,
        factor=2,
        max_backoff_ms=300,
    )
    async def get_with_retry(self, key: MemKey) -> bytes:
        """
        The reason for retrying GET calls on DynamoDB is to do with its
        eventual consistency guarantees. An immediate read-after-write
        may fail but, later, succeed when the write is propagated across
        all storage nodes. We can afford to do this since our use-case
        of Dynamo is immutable & append-only, so when a read returns
        "Not Found" for a key that was recently written, it is essentially
        seeing a stale state, and should, therefore, be retried.

        Backoff configurations match AWS SDK defaults found here
        https://github.com/aws/aws-sdk-java/blob/dec8dfea84dc9433aacb82d27c3ac0def9e04d17/aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java#L29
        """
        return await self.get(key)

    async def set(self, key: MemKey, value: bytes) -> None:
        await self.client.put_item(
            TableName=self.table_name, Item={**self.key(key), "value": {"B": value}}
        )

    async def delete(self, key: MemKey) -> None:
        await self.client.delete_item(
            TableName=self.table_name,
            Key=self.key(key),
        )

    async def set_new_value(self, key: MemKey, value: bytes) -> None:
        """Set a value, ensuring none was previously set"""
        try:
            await self.client.update_item(
                TableName=self.table_name,
                Key=self.key(key),
                UpdateExpression="SET #value = :value",
                ExpressionAttributeNames={"#value": "value"},
                ExpressionAttributeValues={":value": {"B": value}},
                ConditionExpression="attribute_not_exists(#value)",
            )
        except self.client.exceptions.ConditionalCheckFailedException as e:
            raise CompareMismatch() from e

    async def compare_and_set(self, key: MemKey, value: bytes, expected: bytes) -> None:
        if expected is None:
            raise ValueError("dynamo cannot CAS a missing value")
        try:
            await self.client.update_item(
                TableName=self.table_name,
                Key=self.key(key),
                UpdateExpression="SET #value = :value",
                ExpressionAttributeNames={"#value": "value"},
                ExpressionAttributeValues={
                    ":value": {"B": value},
                    ":expected": {"B": expected},
                },
                ConditionExpression="#value = :expected",
            )
        except self.client.exceptions.ConditionalCheckFailedException as e:
            raise CompareMismatch() from e

    async def compare_and_delete(self, key: MemKey, expected: bytes) -> None:
        if expected is None:
            raise ValueError("dynamo cannot CAS delete a missing value")
        try:
            await self.client.delete_item(
                TableName=self.table_name,
                Key=self.key(key),
                ConditionExpression="attribute_exists(#value) AND #value = :expected",
                # value is a reserved word in DynamoDB
                ExpressionAttributeNames={"#value": "value"},
                ExpressionAttributeValues={":expected": {"B": expected}},
            )
        except self.client.exceptions.ConditionalCheckFailedException as e:
            raise CompareMismatch() from e

    async def create_table(self) -> None:
        try:
            await self.client.create_table(
                TableName=self.table_name,
                KeySchema=[
                    {"AttributeName": "pk", "KeyType": "HASH"},
                    {"AttributeName": "sk", "KeyType": "RANGE"},
                ],
                AttributeDefinitions=[
                    {"AttributeName": "pk", "AttributeType": "S"},
                    {"AttributeName": "sk", "AttributeType": "S"},
                ],
                # TODO make this configurable? Should this method even exist?
                ProvisionedThroughput={"ReadCapacityUnits": 5, "WriteCapacityUnits": 5},
            )
        except self.client.exceptions.ResourceInUseException:
            pass
